### PR TITLE
forbid aliasing between the outputs of a differentiable graph

### PR DESCRIPTION
**Pull request by Krovatkin**
Fixes #{issue number}

### CODE REVIEW

File: test/jit/test_autodiff_subgraph_slicing.py [hunk 1]
```diff
@@ -372,3 +372,125 @@ def fn(x, k, cond):
...
+        %0 : int[] = prim::Constant[value=[2, 2, 1]]()
+        %1 : int = prim::Constant[value=0]()
+        %2 : Tensor[] = aten::split_with_sizes(%b, %0, %1)
+        %3 : (Tensor[], Tensor[]) = prim::TupleConstruct(%b, %2)
+        return (%3)
+"""
+
+            graph = torch._C.parse_ir(input_str)
+            torch._C._jit_pass_create_autodiff_subgraphs(graph, 1)
+            FileCheck().check("Tensor = prim::DifferentiableGraph") \
```
Comment (eellison) [inline test/jit/test_autodiff_subgraph_slicing.py:415]: can we test the relu remains?

File: torch/csrc/jit/passes/create_autodiff_subgraphs.cpp [hunk 2]
```diff
@@ -53,20 +67,31 @@ class SubgraphSlicer {
...
-        // redundant prim::Constants). Run CSE to clean them up.
-        EliminateCommonSubexpression(curNode->g(attr::Subgraph));
-
-        if (!inlineIfTooSmall(curNode)) {
-          diff_nodes_.push_back(curNode);
-        }
+
+    bool any_changed = true;
+
+    while(any_changed) {
```
Comment (eellison) [inline torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:73]: Can you move the any_changed loop inside unfuseAliasedOutputs? we have no reason to run the other passes in a loop

File: torch/csrc/jit/passes/create_autodiff_subgraphs.cpp [hunk 3]
```diff
@@ -120,6 +145,192 @@ class SubgraphSlicer {
...
+  std::vector<std::set<Value*, topo_cmp_value>> buildAliasedSets(std::shared_ptr<Graph> subgraph) {
+    auto outputs = subgraph->outputs();
+    AliasDb alias_db(subgraph);
+    TORCH_INTERNAL_ASSERT(outputs.size() > 1);
+    std::vector<std::set<Value*, topo_cmp_value>> res;
+    for (auto o : outputs) {
+      auto grouped = false;
+      for (auto& s : res) {
+        auto os = *s.begin();
+        GRAPH_DEBUG("comparing %", o->debugName(), " with %", os->debugName(), " result ", (alias_db.mayContainAlias(os, o)));
```
Comment (eellison) [inline torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:320]: we shouldn't be running `alias_db.mayContainAlias(os, o))` within the debug, we're effectively running it twice now

File: torch/csrc/jit/passes/create_autodiff_subgraphs.cpp [hunk 4]
```diff
@@ -120,6 +145,192 @@ class SubgraphSlicer {
...
+    }
+
+    auto subgraph = subgraphNode->g(attr::Subgraph);
+    GRAPH_DUMP("unfuseAliasedOutputs Subgraph ", subgraph);
+    auto sets = buildAliasedSets(subgraph);
+    GRAPH_DEBUG("buildAliasedSets sets.size() = ", sets.size());
+
+    std::set<Node*, topo_cmp_node> nodes;
+
+    for (auto i : c10::irange(sets.size())) {
```
Comment (eellison) [inline torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:163]: nice!!

File: torch/csrc/jit/passes/create_autodiff_subgraphs.cpp [hunk 5]
```diff
@@ -120,6 +145,192 @@ class SubgraphSlicer {
...
+    s.insert(start);
+
+    for (auto o : start->outputs()) {
+      for (auto use : o->uses()) {
+        collectNodesToUnfuse(use.user, s);
+      }
+    }
+  }
+
+  void unfuseNode(Node* n, Node* subgraphNode) {
```
Comment (eellison) [inline torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:214]: should this be in subgraphUtils ?

File: torch/csrc/jit/passes/create_autodiff_subgraphs.cpp [hunk 6]
```diff
@@ -120,6 +145,192 @@ class SubgraphSlicer {
...
+    subgraphNode->owningGraph()->insertNode(copy);
+
+    for (auto it = output_indices.rbegin(); it != output_indices.rend(); it++) {
+      auto replace_val = local_map[subgraph->outputs().at(*it)];
+      subgraphNode->outputs().at(*it)->replaceAllUsesWith(replace_val);
+      subgraphNode->eraseOutput(*it);
+      subgraph->eraseOutput(*it);
+    }
+
+    // size_t num_outputs = subgraph->outputs().size();
```
Comment (eellison) [inline torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:302]: commented out code

File: torch/csrc/jit/passes/utils/subgraph_utils.h [hunk 7]
```diff
@@ -46,6 +46,10 @@ TORCH_API void mergeNodeIntoSubgraphAndUpdateAliasing(
     Node* subgraphNode,
     AliasDb& db);
 
+TORCH_API std::vector<Node*> unmergeAliasedOutputs(
```
Comment (eellison) [inline torch/csrc/jit/passes/utils/subgraph_utils.h:49]: if we're defining the api in the header here, can we move the rest of the impl to the subgraph_utils.cpp file ?

**Review by eellison — COMMENTED** [review]
nice, looks great !! a few small comments.

cc @jjsjann123 could be relevant for fusing expand etc

File: torch/csrc/jit/passes/create_autodiff_subgraphs.cpp [hunk 8]
```diff
@@ -120,6 +123,37 @@ class SubgraphSlicer {
...
+      // nodes that might get unfused after the current
+      // prim::DifferentiableGraph
+      for (auto n : b->nodes().reverse()) {
+        if (n->kind() == prim::DifferentiableGraph) {
+          // aliased outputs in DifferentiableGraphs must be unfused
+          // since autodiff doesn't know how to handle them correctly
+          // N.B. Note, |= since we don't want `unfuseAliasedOutputs`
+          // to short-circuit
+          any_changed |= SubgraphUtils::unmergeAliasedOutputs(n);
+          any_changed |= SubgraphUtils::unmergeOutputsAilisingInputs(n);
```
Comment (eellison) [inline torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:140]: typo: `Ailising` -> `aliasing`

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 9]
```diff
@@ -362,6 +432,174 @@ Node* createSingletonSubgraphAndUpdateAliasing(
...
+
+bool unmergeAliasedOutputs(Node* subgraphNode) {
+  GRAPH_DEBUG("unfuseAliasedOutputs on ", getHeader(subgraphNode));
+  if (subgraphNode->outputs().size() < 2) {
+    return false;
+  }
+
+  auto subgraph = subgraphNode->g(attr::Subgraph);
+  GRAPH_DUMP("unfuseAliasedOutputs Subgraph ", subgraph);
+  auto sets = buildAliasedSets(subgraph);
```
Comment (eellison) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:463]: would be great if we can only rebuild this in the case that we've merged out a list of tensors, but i guess in a follow up.

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp [hunk 10]
```diff
@@ -136,6 +139,73 @@ void mergeSubgraph(Node* mergeTo, Node* mergeFrom) {
...
+    for (auto use : o->uses()) {
+      collectNodesToUnfuse(use.user, s);
+    }
+  }
+}
+
+std::vector<std::set<Value*, topo_cmp_value>> buildAliasedSets(
+    std::shared_ptr<Graph> subgraph) {
+  auto outputs = subgraph->outputs();
+  AliasDb alias_db(subgraph);
```
Comment (eellison) [inline torch/csrc/jit/passes/utils/subgraph_utils.cpp:181]: in a follow up let's figure out a way to not rebuild this every time

**Review by eellison — APPROVED** [review]
Nice!! in follow ups let's add subgraph utils and find a way to rebuild alias db more minimally

File: torch/csrc/jit/passes/utils/subgraph_utils.cpp (hunk 9 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/utils/subgraph_utils.cpp:463]: agreed, I'll try following up ASAP on this.

**Pull request closed without merging**